ci2/ci/github.py

```diff
@@ -166,6 +180,30 @@ def config(self):
             'sha': self.sha
         }
 
+    def should_build(self):
+        return self.batch is None and (
+            self.most_recent_build is None or self.source_sha != self.most_recent_build.source_sha)
```

Comment:
I think should_build should only be called if batch is None.  I'd change the condition to an assert.